beamer/agent/util.py

```diff
@@ -1,13 +1,42 @@
 from dataclasses import dataclass
+from typing import cast
 
 from eth_utils import is_checksum_address, to_checksum_address
+from web3 import HTTPProvider, Web3
+from web3.contract import Contract
 
-from beamer.typing import ChainId, ChecksumAddress
-
+from beamer.typing import URL, ChainId, ChecksumAddress
 
 _Token = tuple[ChainId, ChecksumAddress]
 
 
+@dataclass
+class BaseChain:
+    w3: Web3
+    id: ChainId
+
+    @property
+    def rpc_url(self) -> URL:
+        provider = cast(HTTPProvider, self.w3.provider)
+        assert provider.endpoint_uri is not None
+        return URL(provider.endpoint_uri)
+
+
+@dataclass
+class Chain(BaseChain):
+    name: str
+    tokens: list[tuple[ChainId, ChecksumAddress]]
+    _contracts: dict[str, Contract]
```

Comment:
> I believe I did this, because it was never accessed outside, and actually shouldn't.

Okay, we still have one external access to this field, namely during instance construction (see my comment above).
We should fix that.

Reply:
I guess @istankovic right here, with a normal class and setter/getter approach we could have a private contracts, but with current setup it is not ideal, so will change this now.